src/block_cache/block_mgr.c

```diff
@@ -641,14 +682,84 @@ __bm_switch_object_readonly(WT_BM *bm, WT_SESSION_IMPL *session, uint32_t object
     return (__bm_readonly(bm, session));
 }
 
+/*
+ * __bm_switch_object_complete --
+ *     Complete switching the active handle to a different object.
+ */
+static int
+__bm_switch_object_complete(WT_BM *bm, WT_SESSION_IMPL *session, uint32_t objectid)
+{
+    bm->objectid_complete = objectid;
+
+    return (__wt_blkcache_sweep_handles(session, bm));
+}
+
+/*
+ * __bm_switch_object_complete_readonly --
+ *     Complete switching the tiered object; readonly version.
+ */
+static int
+__bm_switch_object_complete_readonly(WT_BM *bm, WT_SESSION_IMPL *session, uint32_t objectid)
+{
+    WT_UNUSED(objectid);
+
+    return (__bm_readonly(bm, session));
+}
+
 /*
  * __bm_sync --
  *     Flush a file to disk.
  */
 static int
-__bm_sync(WT_BM *bm, WT_SESSION_IMPL *session, bool block)
+__bm_sync(WT_BM *bm, WT_SESSION_IMPL *session, bool do_block)
 {
-    return (__wt_fsync(session, bm->block->fh, block));
+    if (bm->is_multi_handle)
+        WT_RET(__bm_sync_tiered_handles(bm, session));
+
+    return (__wt_fsync(session, bm->block->fh, do_block));
+}
+
+/*
+ * __wt_blkcache_sweep_handles --
+ *     Free blocks from the manager's handle array if possible.
+ */
+int
+__wt_blkcache_sweep_handles(WT_SESSION_IMPL *session, WT_BM *bm)
+{
+    WT_BLOCK *block;
+    WT_DECL_RET;
+    size_t nbytes;
+    u_int i;
+
+    WT_ASSERT(session, bm->is_multi_handle);
+
+    /*
+     * This function may be called when the reader count for a block has been observed at zero. Grab
+     * the lock and check again to see if we can remove any block from our list. If the count for a
+     * block was zero and is now not zero, because other readers got references in the meantime, the
+     * last of those readers will have another chance to free it.
+     */
+    __wt_writelock(session, &bm->handle_array_lock);
+    for (i = 0; i < bm->handle_array_next; ++i) {
+        block = bm->handle_array[i];
+        if (block->read_count == 0 && __wt_block_eligible_for_sweep(bm, block)) {
+            /* We cannot close the active handle. */
+            WT_ASSERT(session, block != bm->block);
+            WT_TRET(__wt_bm_close_block(session, block));
+
+            /*
+             * To fill the hole just created, shift the rest of the array down. Adjust the loop
+             * index so we'll continue just where we left off.
+             */
+            nbytes = (bm->handle_array_next - i - 1) * sizeof(bm->handle_array[0]);
+            memmove(&bm->handle_array[i], &bm->handle_array[i + 1], nbytes);
```

Comment:
> How big do we expect this array to be?
Hundreds of entries definitely, possibly thousands. It's easy to imagine a workload that steadily inserts things over time and does reads all over the table. If flush once per hour we could wind up with 720 objects per month.

I expect this array is eventually going to have to be a hash table, which will eliminate the need for shuffling array elements around. So maybe it's OK to for now? 

I think there's a backlog ticket for the array -> hash table switch. If not, I'll create one.

Reply:
I see @keitharnoldsmith created WT-12028. Given the existing precedent it may be easier to have the blocks be a linked list instead of an array to be just like all the other list+hash users.

Reply:
After exploring the linked list to an array.  The problem is that any WT_BLOCK (which would be a member of the list) could appear in the set of blocks used by multiple WT_BMs.  That is, we could have a block manager for a "regular" version of a table:, and a different block manager used by a checkpoint.  Without a level of indirection (like "WT_BLOCK_LIST_ELEMENT"), there's no real way for a WT_BLOCK to appear on an arbitrary number of lists.  I could have created the WT_BLOCK_LIST_ELEMENT, but it makes more sense to defer this until WT-12028, and see if it makes sense to do so there.

Reply:
One minor comment here then: Can you add a FIXME-WT-12028 if appropriate and a comment about the potential for the memmove to be slow spot?